### ISSUE TITLE
Query slow on Update 

### ISSUE THREAD

**Issue by de3** [arangodb/arangodb#2297]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [ ] 2.8
- [x] 3.0
- [ ] self-compiled devel branch

On this operating system:
- [ ] DCOS on
  - [ ] AWS
  - [ ] Azure
  - [ ] own infrastructure
- [ ] Linux 
  - [ ] Debian .deb
  - [x] Ubuntu .deb
  - [ ] SUSE   .rpm
  - [ ] RedHat .rpm
  - [ ] Fedora .rpm
  - [ ] Gentoo
  - [ ] docker - official docker library
  - [ ] other:
- [ ] Windows, version: 
- [ ] MacOS, version:

Slow query only happen when `UPDATE` in some collection.
My update is simple:

```
FOR user IN @@collection
FILTER user._key == @key
UPDATE user WITH @patchData IN @@collection
```
took: 28.593268

Total Rows: ~5000
Total Size: ~33MB

if I try repeat same query:

1. took: 14.527 s
2. took: 18.125 s
3. took: 2.041 s

Last iteration is good.

```
FOR doc IN @@collection
FILTER doc._key == @key
UPDATE doc WITH {
    stats: {
        comments: doc.stats.comments + 1
    }
}
IN @@collection
```
took: 27.242827

Total Rows: ~10k
Total Size: ~65MB

I'm not using Transaction.

Another Issue,
I have collection with ~16mil rows, if my `SELECT` query running slow in that collection, it will make `INSERT` query in that collection become slow. It like locking.

Btw, I use Ubuntu 14.04 and my memory size >200GB

**Comment by dothebart** [comment]
You should use `db._explain()` to get an insight of what queries are actualy doing. 
Using indices for `FILTER`s is almost every time the problem.

**Comment by de3** [comment]
@dothebart but in my first query, I just use `_key` for filter, which means it use attributes that already indexed.
this is output from explain.. I use editor query from web UI

```
Query string:
 FOR user IN @@collection
 FILTER user._key == @key
 UPDATE user WITH @patchData IN @@collection

Execution plan:
 Id   NodeType          Est.   Comment
  1   SingletonNode        1   * ROOT
  5   CalculationNode      1     - LET #5 = { "name" : "asd" }   /* json expression */   /* const assignment */
  7   IndexNode            1     - FOR user IN v1_users   /* primary index scan */
  6   UpdateNode           0       - UPDATE user WITH #5 IN v1_users

Indexes used:
 By   Type      Collection   Unique   Sparse   Selectivity   Fields       Ranges
  7   primary   v1_users     true     false       100.00 %   [ `_key` ]   (user.`_key` == "36955328")

Optimization rules applied:
 Id   RuleName
  1   move-calculations-up
  2   remove-data-modification-out-variables
  3   use-indexes
  4   remove-filter-covered-by-index
  5   remove-unnecessary-calculations-2

Write query options:
 Option                   Value
 ignoreErrors             false
 waitForSync              false
 nullMeansRemove          false
 mergeObjects             true
 ignoreDocumentNotFound   false
 readCompleteInput        true

```
And I think problem occurs randomly .. now I get same issue with INSERT and UPDATE query in another collection .. 
do you think my version has bug so I need to upgrade it?

**Comment by dothebart** [comment]
Sorry I did not come back to you on this one. 

The cause of the problem is, that the mmfiles storage engine in arangodb needs a write lock on the collection to update the value. 
While read locks are still in place, the write lock for the `UPDATE` can't be permitted. The write lock queues up for all existing readlocks to clear up, and will once it acquired the write lock update the value. 

So if the read pressure is high on a collection, writes will become slower with the MMFiles storage engine. 

The rocksdb storage engine available with ArangoDB 3.2 brings the cure to this behaviour. 

I will flag this issue apropriately and close it.
